--- pango.orig/pango/pango-layout.c	(revisión: 2401)
+++ pango-1.16.4/pango/pango-layout.c	(revisión: 2402)
@@ -3891,23 +3891,14 @@
 	  int pos;
 	  int char_index;
 
-	  char_index = run->item->offset;
+	  pango_glyph_string_x_to_index (run->glyphs,
+					 layout->text + run->item->offset, run->item->length,
+					 &run->item->analysis,
+					 x_pos - start_pos,
+					 &pos, &char_trailing);
 
-	  if (properties.shape_set)
-	    {
-	      char_trailing = FALSE;
-	    }
-	  else
-	    {
-	      pango_glyph_string_x_to_index (run->glyphs,
-					     layout->text + run->item->offset, run->item->length,
-					     &run->item->analysis,
-					     x_pos - start_pos,
-					     &pos, &char_trailing);
+	  char_index = run->item->offset + pos;
 
-	      char_index += pos;
-	    }
-
 	  /* Convert from characters to graphemes */
 
 	  offset = g_utf8_pointer_to_offset (layout->text, layout->text + char_index);
